### ISSUE TITLE
Try generating the specialized interpreter from `mit --profile` instead of `mit --predictor`

### ISSUE THREAD

**Issue by apt1002** [rrthomas/smite#275]
I think it should be possible to make `gen-labels` consume the file generated by `mit --profile` instead of the file generated by `mit --predictor`. If so, we would obtain some advantages:

 - We could delete all the code for `mit --predictor`.
 - We can try iterating the profile/specialize step. With each iteration, the profile data would become more specialized, and consequently so would the interpreter. This removes the requirement for a fancy history hash function that tries to be a generic solution for any program.
 - We currently generate specializer states for phantom instruction sequences that don't actually occur, because of hash collisions in the predictor. We might be able to avoid that.

I haven't worked out all the details yet. I think we would have to collect a bit more information about the instructions executed by the fallback interpreter, but otherwise I think the information already in the profile would be sufficient to do a good job.

**Comment by apt1002** [comment]
We're starting with an experiment. Branch `generate-specializer-from-profile-275` has a file `src/features/simulate-jit` which generates a plausible (but fictional) stream of instructions and applies an algorithm to it to decide what a JIT compiler might bother to compile.

This experiment is more ambitious than the original goal, in that it attempts a just-in-time solution, but it is also less ambitious in that it makes use of dynamic counters instead of trying to do everything statically.

Currently, the experimental algorithm is missing some important features. It does not implement the "multi-guess" or "fewer states" optimizations that are already working in the specialized interpreter. We hope to remedy this.

**Comment by apt1002** [comment]
In order to reduce the number of states generated, as for #257 ("fewer states"), the default destination for a label when there's no correct guess, instead of simply being its suffix parent, is the first suffix ancestor with at least two suffix children.

Currently we can generate new Labels when the language is both prefix- and suffix-closed. We say a label is "compiled" when it has been visited at least `COMPILE_THRESHOLD` times, and we construct a new label when its prefix parent and suffix parent have both been compiled.

In order to implement #257, we want to change this strategy. Construct a new label when its prefix parent has been compiled, but in place of its suffix parent we use suffix ancestors, chosen as above.

However, we now sometimes find that a Label we are creating lands between an existing node and its existing suffix ancestor, so when we create a Label we must find and connect it between any existing suffix ancestor and its existing suffix child.

**Comment by apt1002** [comment]
Currently we can generate new Labels when the language is both prefix- and suffix-closed. We say a label is "compiled" when it has been visited at least `COMPILE_THRESHOLD` times, and we construct a new label when its prefix parent and suffix parent have both been compiled.

In order to implement #257, we want to change this strategy. Construct a new label when its prefix parent has been compiled, but in place of its suffix parent we use suffix ancestors.

**Comment by apt1002** [comment]
This seems to be working now. I've deleted the old "gen-labels" from the feature branch, as it is no longer needed. We regressed #125 (as explained there) but we have roughly recovered the performance just by generating a better labels file. Hopefully we can now get faster by redoing #125.

Here are some timings for building pForth using "mit -O" in various ways on @apt1002's Sandy Bridge desktop machine. For comparison, the old "mit -O" (current master) takes 15.6s, and "mit" takes 68s. Each row represents a different size of interpreter (controlled using the `--labels` switch of `simulate-jit`). Each column represents a build/profile iteration, with the left-hand column being an interpreter "specialized" for the empty profile, and each subsequent column being an interpreter specialized for the profile made by the previous interpreter.

```
--labels 100	54.2	41.2	39.5	32.8	35.1	31.3	30.2	30.7	34.1
--labels 200	54.3	39.7	20.4	18.7	17.7	17.7	18.7
--labels 400	54.2	42.3	22.8	18.5	17.5	15.8	15.6	14.6	16.4	15.0	16.4
--labels 800	54.3	37.7	18.8	19.0	17.9	19.9	18.2	19.4
```
